python/dgl/node_flow.py

```diff
@@ -0,0 +1,845 @@
+
+        Returns
+        -------
+        Tensor
+            The parent edge id array.
+        """
+        block_id = self._get_block_id(block_id)
+        start = self._block_offsets[block_id]
+        end = self._block_offsets[block_id + 1]
+        return self._edge_mapping.tousertensor()[start:end]
+
+    def set_n_initializer(self, initializer, layer_id=ALL, field=None):
+        """Set the initializer for empty node features.
+
+        Initializer is a callable that returns a tensor given the shape, data type
+        and device context.
+
+        When a subset of the nodes are assigned a new feature, initializer is
+        used to create feature for rest of the nodes.
+
+        Parameters
+        ----------
+        initializer : callable
+            The initializer.
+        layer_id : int
+            the layer to set the initializer.
+        field : str, optional
+            The feature field name. Default is set an initializer for all the
+            feature fields.
+        """
+        if is_all(layer_id):
+            for i in range(self.num_layers):
+                self._node_frames[i].set_initializer(initializer, field)
+        else:
+            self._node_frames[i].set_initializer(initializer, field)
+
+    def set_e_initializer(self, initializer, block_id=ALL, field=None):
+        """Set the initializer for empty edge features.
+
+        Initializer is a callable that returns a tensor given the shape, data
+        type and device context.
+
+        When a subset of the edges are assigned a new feature, initializer is
+        used to create feature for rest of the edges.
+
+        Parameters
+        ----------
+        initializer : callable
+            The initializer.
+        block_id : int
+            the block to set the initializer.
+        field : str, optional
+            The feature field name. Default is set an initializer for all the
+            feature fields.
+        """
+        if is_all(block_id):
+            for i in range(self.num_blocks):
+                self._edge_frames[i].set_initializer(initializer, field)
+        else:
+            self._edge_frames[block_id].set_initializer(initializer, field)
+
+
+    def register_message_func(self, func, block_id=ALL):
+        """Register global message function for a block.
+
+        Once registered, ``func`` will be used as the default
+        message function in message passing operations, including
+        :func:`block_compute`, :func:`prop_flow`.
+
+        Parameters
+        ----------
+        func : callable
+            Message function on the edge. The function should be
+            an :mod:`Edge UDF <dgl.udf>`.
+        block_id : int or ALL
+            the block to register the message function.
+        """
+        if is_all(block_id):
+            self._message_funcs = [func] * self.num_blocks
+        else:
+            self._message_funcs[block_id] = func
+
+    def register_reduce_func(self, func, block_id=ALL):
```

Comment:
Does this means no reduce function can be registered for the first layer?

Reply:
all registered UDFs are for blocks. not sure why reduce should be special.